scripts/dvm

```diff
@@ -469,35 +510,109 @@ dvm() {
   shift
   case $cmd in
     alias)
-      dvm_alias "$@"
+      _dvm_alias "$@"
       ;;
     implode)
-      dvm_implode "$@"
+      _dvm_implode "$@"
       ;;
     install)
-      dvm_install "$@"
+      _dvm_install "$@"
       ;;
     list)
-      dvm_list "$@"
+      _dvm_list "$@"
       ;;
     listall)
-      dvm_listall "$@"
+      _dvm_listall "$@"
       ;;
     upgrade)
-      dvm_upgrade "$@"
+      _dvm_upgrade "$@"
       ;;
     use)
-      dvm_use "$@"
+      _dvm_use "$@"
       ;;
     version)
-      dvm_version "$@"
+      _dvm_version "$@"
       ;;
     *)
       _dvm_usage
       ;;
   esac
 }
 
+
+_dvm_completions() {
+  local CURRENT_WORD="${COMP_WORDS[COMP_CWORD]}"
+  
+  if [[ "$COMP_CWORD" == 0 ]]; then
+    COMPREPLY=($(compgen -W "dvm" -- "$CURRENT_WORD"))
+    return
+  fi
+  if [[ "$COMP_CWORD" == 1 ]]; then
+    COMPREPLY=($(compgen -W "alias implode install list listall upgrade use version" -- "$CURRENT_WORD"))
+    return
+  fi
+  
+  case "${COMP_WORDS[1]}" in
+    # dvm alias
+    alias)
+      if [[ "$COMP_CWORD" == 2 ]]; then
+        COMPREPLY=($(compgen -W "create update delete list" -- "$CURRENT_WORD"))
+      fi
+      ;;
+    
+    # dvm implode
+    implode)
+      ;;
+    
+    # dvm install
+    install)
+      if [[ "$COMP_CWORD" == 2 ]]; then
+        COMPREPLY=($(compgen -W "$(_dvm_list_repo 'stable')" -- "$CURRENT_WORD"))
+      fi
+      ;;
+    
+    # dvm list
+    list)
+      ;;
+    
+    # dvm listall
+    listall)
+      if [[ "$COMP_CWORD" == 2 ]]; then
+        COMPREPLY=($(compgen -W "--dev --beta" -- "$CURRENT_WORD"))
+      fi
+      ;;
+    
+    # dvm upgrade
+    upgrade)
+      ;;
+    
+    # dvm use
+    use)
+      if [[ "$COMP_CWORD" == 2 ]]; then
+        COMPREPLY=($(compgen -W "$(_dvm_list) --default" -- "$CURRENT_WORD"))
+      fi
+      if [[ "$COMP_CWORD" == 3 ]]; then
+        if [[ "${COMP_WORDS[2]}" == '--default' ]]; then
+          COMPREPLY=($(compgen -W "$(_dvm_list)" -- "$CURRENT_WORD"))
+        else
+          COMPREPLY=($(compgen -W "--default" -- "$CURRENT_WORD"))
+        fi
+      fi
+      ;; 
+    
+    # dvm version
+    version)
+      ;;
+
+    # Others don't work!
+    *)
+      ;;
+  esac
+}
+
+# Setup bash completion
+complete -F _dvm_completions dvm
```

Comment:
We should probably guard this behind a conditional so as not to cause errors for anyone using other shells. Maybe something like:

```shell
if [[ "$(basename $SHELL)" == "bash" ]]; then
  complete -F _dvm_completions dvm
fi
```

Reply:
Actually, come to think of it, another alternative would be to break out a separate completions file; that would allow us to keep the shell-specific code separate.

Reply:
That's a good idea! :D

Reply:
Note: stackoverflow suggested using `BASH_VERSION` instead of `$SHELL` because `$SHELL` is the users default shell, not necessarily the shell currently running.